runtime/include/MemRefUtils.hpp

```diff
@@ -0,0 +1,143 @@
+    T *data_allocated;
+    T *data_aligned;
+    size_t offset;
+    size_t sizes[R];
+    size_t strides[R];
+};
+
+/**
+ * A multi-dimensional view for MemRef<T, R> types.
+ *
+ * @tparam T The underlying data type
+ * @tparam R The Rank (R >= 0)
+ *
+ * @note A forward iterator is implemented in this view for traversing over the entire
+ * elements of MemRef types rank-by-rank starting from the last dimension (R-1). For example,
+ * The MemRefView iterator for MemRef<T, 2> starts from index (0, 0) and traverses elements
+ * in the following order:
+ * (0, 0), ..., (0, sizes[1]-1), (1, 0), ..., (1, sizes[1]-1), ... (sizes[0]-1, sizes[1]-1).
+ */
+template <typename T, size_t R> class MemRefView {
+  private:
+    const MemRefT<T, R> *buffer;
+    size_t tsize; // total size
+
+  public:
+    class MemRefIter {
+      private:
+        const MemRefT<T, R> *buffer;
+        int64_t loc; // physical index
+        size_t indices[R] = {0};
+
+      public:
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = T;
+        using difference_type = std::ptrdiff_t;
+        using pointer = T *;
+        using reference = T &;
+
+        MemRefIter(const MemRefT<T, R> *_buffer, int64_t begin_idx)
+            : buffer(_buffer), loc(begin_idx)
+        {
+        }
+        pointer operator->() const { return &buffer->data_aligned[loc]; }
+        reference operator*() const { return buffer->data_aligned[loc]; }
+        MemRefIter &operator++()
+        {
+            int64_t next_axis = -1;
+            int64_t idx;
+            for (int64_t i = R; i > 0; --i) {
+                idx = i - 1;
+                if (indices[idx]++ < buffer->sizes[idx] - 1) {
+                    next_axis = idx;
+                    break;
+                }
+                indices[idx] = 0;
+                loc -= (buffer->sizes[idx] - 1) * buffer->strides[idx];
+            }
+
+            loc = next_axis == -1 ? -1 : loc + buffer->strides[next_axis];
+            return *this;
+        }
+        MemRefIter operator++(int)
+        {
+            auto tmp = *this;
+            int64_t next_axis = -1;
+            int64_t idx;
+            for (int64_t i = R; i > 0; --i) {
+                idx = i - 1;
+                if (indices[idx]++ < buffer->sizes[idx] - 1) {
+                    next_axis = idx;
+                    break;
+                }
+                indices[idx] = 0;
+                loc -= (buffer->sizes[idx] - 1) * buffer->strides[idx];
+            }
+
+            loc = next_axis == -1 ? -1 : loc + buffer->strides[next_axis];
+            return tmp;
+        }
+        bool operator==(const MemRefIter &other) const
+        {
+            return (loc == other.loc && buffer == other.buffer);
+        }
+        bool operator!=(const MemRefIter &other) const { return !(*this == other); }
+    };
+
+    explicit MemRefView(const MemRefT<T, R> *_buffer, size_t _size) : buffer(_buffer), tsize(_size)
```

Comment:
The total size should be computable from the MemRef descriptor itself, so we don't need to require the caller to do it :)

Reply:
Good catch! Part of the old design :)